### ISSUE TITLE
Issues with detecting and avoiding duplicates

### ISSUE THREAD

**Issue by krokofant** [spaam/svtplay-dl#1343]
### svtplay-dl versions:
3.2

### Operating system and Python version:
Windows 10 (2004)
Python 3.8 (included in release)
Python 3.9.2 (global install)

### What is the issue:
There doesn't seem to exist a consistent way avoiding to download duplicates, especially when the episode IDs might change.

#### When id's has changed
`svtplay-dl.exe --subtitle --all-episodes https://www.svtplay.se/exit` when _IDs have not changed_ will skip the already downloaded episodes unless `-force` is specified

When id's _have changed_ it will download the episode again.

#### Downloading only episodes after date
`svtplay-dl.exe --subtitle --after-date 2021-03-06 https://www.svtplay.se/exit` renders this error:
```
ERROR: No videos found. This mode is not supported anymore. Need the url with the video.
```

#### Downloading only latest 5 episodes
`svtplay-dl.exe --subtitle --all-last 5 https://www.svtplay.se/exit` render this error:
```
ERROR: No videos found. This mode is not supported anymore. Need the url with the video.
```

#### (Bad workaround) Downloading only missing episodes with some scripting
1. Find the available episodes with `svtplay-dl.exe --subtitle --all-episodes --get-only-episode-url https://www.svtplay.se/exit`
2. Exclude the local episodes by matching season and episode

The issue with this approach is that the output is 1) hard to parse and 2) is doesn't return the `s0Xe0X` for you so you can't easily match this with your local episodes. Is there an easier way parse the episode information from svtplay-dl?

#### (Better workaround) Download only missing via exclusion with powershell:
1. Get a list of all local episodes and create a list of each s0Xe0X pattern
2. Use the --exclude option to exclude episodes matching this pattern

```powershell
svtplay-dl.exe --subtitle --all-episodes --exclude $((Get-Item * -Filter *.mp4 | ForEach-Object { $_.BaseName -replace '.*(s\d+e\d+).*','$1'}) -join ',') https://www.svtplay.se/exit
```

It works but it's not especially fast.

#### (Future solution?) Custom output format
Can the output format be customized in any way? I'm thinking that specifying a custom output file _format_ could help detecting duplicates when downloading multiple episodes.

Something like:
```
svtplay-dl.exe --all-episodes URL --episode-name-format '%series%.%season%%episode%%title%'
```
then if the file names match a existing local one it'll be skipped, unless `-force` is supplied.

**Comment by spaam** [comment]
Now you are mixing different issues in one.
1) `When id's has changed`  that is not something i can do. it is outside my power.
2) `Downloading only episodes after date`  there is no video on that page, it will work on the page with the video.
3) `Downloading only latest 5 episodes` i guess i need to add more checks that `--all-last` only can be used with ´--all-episodes`  i guess `--help` text section as not enough.
4) `Custom output format` already have that , but its only exposed via config file. i guess you have missed the section in https://svtplay-dl.se/config/ about filename

**Comment by krokofant** [comment]
@spaam I overlooked the config section 🤦‍♂️ I only mentioned issue 1. since it made it harder to avoid duplicates. With the custom format supported via the config, avoiding duplicates based on output filename should be enough.

However, the _custom filename format_ with _all episodes_ option does not seem to skip existing episodes properly. It downloads both audio and video, and _replaces the old file_ instead of skipping. No `--force` specified.

```powershell
INFO: Episode 1 of 11
INFO: Url: http://www.svtplay.se/video/24136938/exit/exit-avsnitt-1
INFO: Outfile: exit - s01e01 - avsnitt.1.srt
WARNING: File (exit - s01e01 - avsnitt.1.srt) already exists. Use --force to overwrite
INFO: Selected to download dash, bitrate: 5700 format: h264-51
INFO: Outfile: exit - s01e01 - avsnitt.1.m4a
[506/506][================================================================================] ETA: 0:00:00
INFO: Outfile: exit - s01e01 - avsnitt.1.mp4
WARNING: File (exit - s01e01 - avsnitt.1.mp4) already exists. Use --force to overwrite
INFO: Merge audio and video into exit - s01e01 - avsnitt.1.mp4
INFO: Merging done, removing old files.
```

#### Replication
`svtplay-dl --config svtplay-dl.yml https://www.svtplay.se/exit`
```yaml
# svtplay-dl.yml
service:
  svtplay:
    all_episodes: true
    subtitle: true
    filename: "{title} - s{season}e{episode} - {episodename}.{ext}"
```

**Comment by spaam** [comment]
interesting bug, yeah it should skip the files :)

**Comment by spaam** [comment]
the issue here is that i excluded the m4a file from double checking if the file exists so when we download the video (mp4) it wont complain  that we already downloaded the file :)

**Comment by pythonuser3856** [comment]
Excluding m4a checking is a good thing since it allows parallel fetching of the m4a and the mp4; it works as long as m4a concludes first.
I use that behavior as a feature of svtplay-dl !

What would be helpful, though, is if it's possible to auto-create file names the way it was done before. That would be useful at sorting out potential dups.
Details in file naming by svtplay-dl that has changed (for most but not all items) are that season and episode "sXXeYY" is missing and that the word delimiter has been changed from a hyphen ("-") to a punctuation (".")

**Comment by spaam** [comment]
yeah its a good thing.  then i looked into it and was running into if we have like file-s01e01.mp4, then for some reason it choose hls to download. it will download it again. :)

before downloading the files is not a terrible idea. big brain idea there, totally forgot about that haha. right now we only do that when we want to exclude files.  thanks for that :D 

´hyphen ("-") to a punctuation (".")`  hmm okey 🤔

**Comment by spaam** [comment]
yeah.  there are issues here and there.  this will hopefully be fixed later this week or next.

**Comment by pythonuser3856** [comment]
Then I suggest to leave it "as is"; current file naming will soon be the new normality.
It's great to have got season and episode numbering added !

**Comment by anttin2020** [comment]
3.5 still has the problem that it always loads the audio fist and then when starting to load video, it detects the already existing file. Please fix this :)

**Comment by anttin2020** [comment]
Same problem continues with 3.6. The audio file gets downloaded every time and it does not even get deleted when it detects the existing video. Please make the video download first and detect existing from that and only if that does not exist, then download the audio.

**Comment by anttin2020** [comment]
It seems this is not a significant problem for anyone else and spaam does not think this is worth fixing?
No comments for a week :o Am I missing something?

**Comment by spaam** [comment]
@anttin2020  i planned to look at this issue this week. i thought crashes while downloading files is far more important than this issue.

**Comment by anttin2020** [comment]
@spaam, yes of course, no worries, I was just wondering why nobody commented so I thought I had missed some trick or something :) Thank you!

**Comment by anttin2020** [comment]
> Same problem continues with 3.6. The audio file gets downloaded every time and it does not even get deleted when it detects the existing video. Please make the video download first and detect existing from that and only if that does not exist, then download the audio.
15 days ago:
> @anttin2020 i planned to look at this issue this week. i thought crashes while downloading files is far more important than this issue.

Are you still struggling with those serious problems or could it be time for this now? ;)

No worries if you are busy, just don't forget this, it is important thouhg not fatal :)

**Comment by spaam** [comment]
i committed thing to master that should fix this. please test it before i make an release :)

**Comment by anttin2020** [comment]
Thank you but I cannot find any new files in Master :o

**Comment by krokofant** [comment]
@anttin2020 You can download the latest pre-release/snapshot from here https://svtplay-dl.se/download/snapshots/20210509_195246/

**Comment by anttin2020** [comment]
Thank you, I didn't notice the snapshot link on download page, sorry :)

But how did you get this to work at all? No matter what URL I try, it always crashes.
With --verbose I get reason:
 File "bin/svtplay-dl", line 7, in <module>
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\__init__.py", line 69, in main
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py", line 60, in get_media
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py", line 92, in get_all_episodes
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py", line 115, in get_one_media
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\text.py", line 60, in exclude
TypeError: argument of type 'WindowsPath' is not iterable

So here's the status for latest snapshots:
OK but always loads the audio: 20210427_174409/                                   27-Apr-2021 17:49                   -
crash: 20210509_001313/                                   09-May-2021 00:16                   -
crash: 20210509_195246/                                   09-May-2021 19:55                   -

**Comment by spaam** [comment]
it seems like you are using python version that is older than 3.8. please upgrade to 3.8 or something newer.

**Comment by anttin2020** [comment]
Ok, here is all info:
command line:
svtplay-dl.exe --verbose --config config.yml https://www.svtplay.se/jakten-pa-en-mordare

config.yml:
default:
    path: "Y:/DL/svtplaydl"
    subtitle: true
    merge_subtitle: true
    all_episodes: true
    exclude: syntolkat
    remux: true
    filename: "{title}.s{season}e{episode}.{episodename}.{ext}"

--verbose result:
svtplay-dl.exe --verbose --config config.yml https://www.svtplay.se/jakten-pa-en-mordare
DEBUG [1620654225.81136] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py/get_media: version: 3.7-8-gdcf0e6f
DEBUG [1620654225.8133838] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\service\__init__.py/__init__: service: svtplay
DEBUG [1620654225.8143528] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\http.py/request: HTTP getting 'https://www.svtplay.se/jakten-pa-en-mordare'
DEBUG [1620654225.8173769] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_new_conn: Starting new HTTPS connection (1): www.svtplay.se:443
DEBUG [1620654225.915084] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_make_request: https://www.svtplay.se:443 "GET /jakten-pa-en-mordare HTTP/1.1" 200 38377
DEBUG [1620654225.9350545] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\service\__init__.py/__init__: service: svtplay
INFO [1620654225.9350545] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py/get_all_episodes: Episode 1 of 13
INFO [1620654225.9350545] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py/get_all_episodes: Url: http://www.svtplay.se/video/29015622/jakten-pa-en-mordare/jakten-pa-en-mordare-avsnitt-1
DEBUG [1620654225.9350545] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\http.py/request: HTTP getting 'http://www.svtplay.se/video/29015622/jakten-pa-en-mordare/jakten-pa-en-mordare-avsnitt-1'
DEBUG [1620654225.9360273] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_new_conn: Starting new HTTP connection (1): www.svtplay.se:80
DEBUG [1620654226.045879] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_make_request: http://www.svtplay.se:80 "GET /video/29015622/jakten-pa-en-mordare/jakten-pa-en-mordare-avsnitt-1 HTTP/1.1" 307 0
DEBUG [1620654226.047845] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_new_conn: Starting new HTTPS connection (1): www.svtplay.se:443
DEBUG [1620654226.120655] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_make_request: https://www.svtplay.se:443 "GET /low-prio/video/29015622/jakten-pa-en-mordare/jakten-pa-en-mordare-avsnitt-1 HTTP/1.1" 200 38873
DEBUG [1620654226.1435933] D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\http.py/request: HTTP getting 'https://api.svt.se/video/KNwz5vr'
DEBUG [1620654226.1465867] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_new_conn: Starting new HTTPS connection (1): api.svt.se:443
DEBUG [1620654226.2485027] C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\urllib3\connectionpool.py/_make_request: https://api.svt.se:443 "GET /video/KNwz5vr HTTP/1.1" 200 1318
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\cx_Freeze\initscripts\__startup__.py", line 81, in run
  File "C:\hostedtoolcache\windows\Python\3.8.9\x64\lib\site-packages\cx_Freeze\initscripts\Console.py", line 36, in run
  File "bin/svtplay-dl", line 7, in <module>
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\__init__.py", line 69, in main
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py", line 60, in get_media
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py", line 92, in get_all_episodes
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\getmedia.py", line 115, in get_one_media
  File "D:\a\svtplay-dl\svtplay-dl\lib\svtplay_dl\utils\text.py", line 60, in exclude
TypeError: argument of type 'WindowsPath' is not iterable

I have tried many URLs, all give the same result, also with -A or without it.

**Comment by krokofant** [comment]
I tested the latest snapshot as of now and it seems far better than before. I'm closing the issue. Thank you @spaam 💙 

```powershell
svtplay-dl 3.7-11-ge6379d9
Python 3.9.4
```

**Comment by spaam** [comment]
@krokofant thanks for testing this new thing <3
